### PR TITLE
fix(appset): regard selector in nested matrix generator's generators

### LINKED ISSUE #11982: selector in nested matrix generators not regarded
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

If using a nested matrix generator, the selectors of the nested generators are not regarded in any way.

**To Reproduce**

Setup:

https://github.com/Roshick/poc-payload
https://github.com/Roshick/poc-workload

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: test-set
spec:
  generators:
  - matrix:
      generators:
      - git:
          repoURL: https://github.com/Roshick/poc-platform.git
          revision: HEAD
          pathParamPrefix: clusterCfg
          files:
          - path: clusters/*.yaml
      - matrix:
          generators:
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: serviceCfg
              files:
              - path: services/*/service.yaml
            selector:
              matchExpressions:
              - {key: providers.test-set-foo, operator: Exists}
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: envCfg
              files:
              - path: services/{{service}}/dev.yaml
  template:
    metadata:
      name: '{{service}}-dev-{{cluster}}'
      labels:
        environment.type: dev
        service: '{{service}}'
    spec:
      project: default
      source:
        repoURL: '{{repoURL}}'
        targetRevision: '{{targetRevision}}'
        chart: '{{chart}}'
      destination:
        server: https://kubernetes.default.svc
        namespace: default
```

**Expected behavior**

As https://github.com/Roshick/poc-workload/blob/main/services/service-a/service.yaml does not contain providers.test-set-foo, the inner git generator in the nested matrix generator should not produce any values and the application should not be created.

Setting the selector to the outer matrix generator produces the expected results:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: test-set
spec:
  generators:
  - matrix:
      generators:
      - git:
          repoURL: https://github.com/Roshick/poc-platform.git
          revision: HEAD
          pathParamPrefix: clusterCfg
          files:
          - path: clusters/*.yaml
      - matrix:
          generators:
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: serviceCfg
              files:
              - path: services/*/service.yaml
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: envCfg
              files:
              - path: services/{{service}}/dev.yaml
        selector:
          matchExpressions:
          - {key: providers.test-set-foo, operator: Exists}
  template:
    metadata:
      name: '{{service}}-dev-{{cluster}}'
      labels:
        environment.type: dev
        service: '{{service}}'
    spec:
      project: default
      source:
        repoURL: '{{repoURL}}'
        targetRevision: '{{targetRevision}}'
        chart: '{{chart}}'
      destination:
        server: https://kubernetes.default.svc
        namespace: default
```

**Version**

2.6.0-rc2

### PR DESCRIPTION
**Pull request by Roshick**
Fixes #11982

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by blakepettersson** [comment]
Does this work as intended? If so, could you write some e2e tests to exercise that functionality?

**Comment by Roshick** [comment]
The fix works, tried with my local cluster. I'll have a look into the e2e testing setup this project uses! Need to check why my branch is failing after the rebase anyway.

**Comment by Roshick** [comment]
@blakepettersson I've added an e2e test that covers the fix.

**Comment by Roshick** [comment]
@crenshaw-dev as discussed I've added a flag and log some warnings if someone is using terminal generators with selectors without having the flag enabled. I'm not sure if this is the cleanest solution (i.e. adding the terminal selector when unmarshalling and removing it afterwards) instead of not adding it in the first place when going from terminal to nested, but the latter would need access to the application set (1. for the flag, 2. for some meaningful log warnings).

if this solution is okay, I would write 3 more e2e tests for the remaining branches (matrix->merge, merge->matrix, merge->merge).

**Comment by Roshick** [comment]
@crenshaw-dev pull request is ready for review now

File: applicationset/generators/matrix.go [hunk 1]
```diff
@@ -168,6 +172,20 @@ func getMatrixGenerator(r argoprojiov1alpha1.ApplicationSetNestedGenerator) (*ar
 	return matrix.ToMatrixGenerator(), nil
 }
 
+func applyNestedMatrixSelectors(appSet *argoprojiov1alpha1.ApplicationSet, matrix *argoprojiov1alpha1.MatrixGenerator) {
```
Comment (crenshaw-dev) [inline applicationset/generators/matrix.go:175]: Can we consolidate this to one function and call it something like `dropDisabledSelectors`?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: applicationset/generators/matrix.go (hunk 1 above)
Comment (Roshick) [reply applicationset/generators/matrix.go]: Done, please check if the solution is okay like that.

**Comment by ishitasequeira** [comment]
@Roshick would you mind resolving conflicts? I will give the PR a review today.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]

**Review by ishitasequeira — CHANGES_REQUESTED** [review]

**Review by ishitasequeira — APPROVED** [review]
LGTM!! @crenshaw-dev could you take another look?

**Comment by Roshick** [comment]
@crenshaw-dev would appreciate another review :)

**Comment by Roshick** [comment]
@crenshaw-dev all fixed up now and ready to be reviewed! :)

**Review by crenshaw-dev — APPROVED** [review]
Thanks, @Roshick!

**Comment by crenshaw-dev** [comment]
Drat, I broke auto-merge. I'll solve merge conflicts and push, then I'll merge after checks pass.

**Pull request merged**